File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
def check(lens,m):
    sum = 0
    for l in lens:
        sum += l // m
    if sum >= n:
        return True
    return False

k, n = map(int,input().split())
lens = [int(input()) for _ in range(k)]
max_len = max(lens)

lo = 0; hi = max_len+1
while lo + 1 != hi:
    mid = (lo+hi)//2
    if check(lens,mid):
        lo = mid
    else:
        hi = mid
print(lo)

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.cc
#include <iostream>
#include <vector>

using namespace std;

int k, n;
vector<int> lans;
int maxNum = 0;

bool Check(int len)
{
    int sum = 0;
    for(int i=0; i<k; i++)
    {
        sum += lans[i]/len;
    }
    if(sum >= n)    return true;
    return false;
}

int Binary(long long s, long long e)
{
    // cout<<s<<' '<<e<<'\n';
    if(s > e)   return e;
    long long mid = (s+e)/2;
    if(Check(mid))  return Binary(mid+1, e);
    else            return Binary(s, mid-1);
}

int main()
{
    cin.tie(nullptr); ios::sync_with_stdio(false);

    cin>>k>>n;
    lans.resize(k);
    for(int i=0; i<k; i++)
    {
        cin>>lans[i];
        if(maxNum < lans[i]) maxNum = lans[i];
    }

    cout<<Binary(1, maxNum)<<endl;
}